src/ast/irbuilderbpf.cpp

```diff
@@ -1669,143 +1644,146 @@ Value *IRBuilderBPF::CreateStrncmp(Value *str1,
-                                { getInt32(i + j) });
-        l = CreateLoad(getInt8Ty(), ptr_l);
-      }
-
-      Value *r;
-      if (literal2)
-        r = getInt8(literal2->c_str()[i]);
-      else {
-        auto *ptr_r = CreateGEP(getInt8Ty(),
-                                CreatePointerCast(val2,
-                                                  getInt8Ty()->getPointerTo()),
-                                { getInt32(i) });
-        r = CreateLoad(getInt8Ty(), ptr_r);
-      }
-
-      Value *cmp_null = CreateICmpEQ(r, null_byte, "strcontains.cmp_null");
-      CreateCondBr(cmp_null, done_true, cmp_char);
-
-      SetInsertPoint(cmp_char);
-
-      Value *cmp = CreateICmpNE(l, r, "strcontains.cmp");
-      CreateCondBr(cmp, first_loop, second_loop);
+  CreateBr(empty_check);
+  SetInsertPoint(empty_check);
+  Value *needle_first = CreateLoad(
+      getInt8Ty(), CreateGEP(getInt8Ty(), needle, { getInt64(0) }));
+  Value *needle_first_null = CreateICmpEQ(needle_first, null_byte);
+  CreateCondBr(needle_first_null, done_true, outer_cond);
+
+  SetInsertPoint(outer_cond);
+  Value *i_val = CreateLoad(getInt64Ty(), i);
+  Value *haystack_inbounds = CreateICmpULT(i_val, getInt64(haystack_sz));
+  Value *haystack_char = CreateLoad(
+      getInt8Ty(), CreateGEP(getInt8Ty(), haystack, { i_val }));
+  Value *haystack_valid = CreateICmpNE(haystack_char, null_byte);
+  Value *haystack_outer_cond = CreateAnd({ haystack_inbounds, haystack_valid });
+  CreateCondBr(haystack_outer_cond, outer_body, done_false);
+
+  SetInsertPoint(outer_body);
+  CreateStore(getInt64(0), j);
+  CreateBr(inner_cond);
+
+  SetInsertPoint(inner_cond);
+  Value *j_val = CreateLoad(getInt64Ty(), j);
+  Value *needle_inbounds = CreateICmpULT(j_val, getInt64(needle_sz));
+  CreateCondBr(needle_inbounds, inner_body, outer_incr);
+
+  SetInsertPoint(inner_body);
+  Value *needle_char = CreateLoad(getInt8Ty(),
+                                  CreateGEP(getInt8Ty(), needle, { j_val }));
+  Value *needle_null = CreateICmpEQ(needle_char, null_byte);
+  CreateCondBr(needle_null, done_true, inner_notnull);
+
+  // Inner conditional matching haystack char with needle char
+  SetInsertPoint(inner_notnull);
+  Value *haystack_cmp_idx = CreateAdd(i_val, j_val);
+  Value *haystack_cmp_outbounds = CreateICmpUGE(haystack_cmp_idx,
+                                                getInt64(haystack_sz));
+  Value *haystack_cmp_char = CreateLoad(
+      getInt8Ty(), CreateGEP(getInt8Ty(), haystack, { haystack_cmp_idx }));
+  Value *haystack_cmp_null = CreateICmpEQ(haystack_cmp_char, null_byte);
+  Value *haystack_cmp_needle = CreateICmpNE(haystack_cmp_char, needle_char);
+  Value *haystack_cmp_cond = CreateOr(
+      { haystack_cmp_outbounds, haystack_cmp_null, haystack_cmp_needle });
+  CreateCondBr(haystack_cmp_cond, outer_incr, inner_incr);
+
+  SetInsertPoint(inner_incr);
+  CreateStore(CreateAdd(CreateLoad(getInt64Ty(), j), getInt64(1)), j);
```

Comment:
Note: this could be replaced by a PHI node to save stack space. Not sure if it's worth the work, though. And maybe LLVM will do it for us.

Reply:
Ah I see https://llvm.org/docs/LangRef.html#id319 . Trying to wrap my head around that example :P

Reply:
Lemme follow up on this in a separate PR.